### PR TITLE
[RFC 0086] NixOS generation APIs

### PR DESCRIPTION
**Pull request by miallo**
[rendered](https://github.com/miallo/rfcs/blob/86/rfcs/0086-nixos-generation-apis.md)

### CODE REVIEW

**Comment by miallo** [comment]
I don't know, if an RFC is the "right" thing, but I think, input from the comunity would be valuable and I didn't get answers to the remaining questions on [discourse](https://discourse.nixos.org/t/what-to-show-in-list-of-nixos-generations/11216).

For not force-pushing but still having a clean patch-set with 3 readable commits instead of the mess of the [PR](https://github.com/NixOS/nixpkgs/pull/105910), I have a [seperate branch](https://github.com/miallo/nixpkgs/tree/nixos-rebuild/clean-commits) for the implementation where the commits are clean. I will keep it updated and clean by force-pushing, so it stays readable and in sync

**Review by Mic92 — APPROVED** [review]
What is not clear to me, is what is causing these incompatibilities? We already have `init-interface-version` for systemd related incompatibilities.

**Comment by Mic92** [comment]
What is not clear to me, is what is causing these incompatibilities? We already have init-interface-version for systemd related incompatibilities.

**Comment by Mic92** [comment]
This looks like a straight forward feature to me. Is there anything in particular that you would want to discuss/solve with having this as a RFC?

File: rfcs/0086-nixos-generation-apis.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+related-issues: https://github.com/NixOS/nixpkgs/pull/105910, https://github.com/NixOS/nixpkgs/issues/24374
+---
+
+# Summary
+[summary]: #summary
+
+On a typical system there are lots of generations. However rolling back to a
+specific one without rebooting is hard and it is even more difficult to get an
+overview of the available generations. This RFC proposes adding
+`nixos-rebuild --generation <number>` for rolling back to a specific generation
```
Comment (zimbatm) [inline rfcs/0086-nixos-generation-apis.md:17]:
```suggestion
`nixos-rebuild switch --generation <number>` for rolling back to a specific generation
```

File: rfcs/0086-nixos-generation-apis.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+[54188](https://github.com/NixOS/nixpkgs/issues/54188).
+Though @edolstra [said it would be okay for now](https://github.com/NixOS/nixpkgs/issues/105910#issuecomment-754036275).
+
+
+## nixos-version
+
+Changing `nixos-version` so that a mistyped parameter throws an error improves
+error detection, but is a potentially breaking change, if there were scripts
+relying on mistyped flags. Though my opinion is, that it is better to throw
+meaningful errors instead of just doing something.
```
Comment (zimbatm) [inline rfcs/0086-nixos-generation-apis.md:141]: This looks like an independent change. But Is worth it nonetheless.

If scripts mistakenly passed an argument then it's up to them to fix their invocation.

File: rfcs/0086-nixos-generation-apis.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+- Default value for configuration revision field if unset? (probably not)
+- Truncate revision (maybe if hash only)?
+- Put the `(current)` tag in it's own column?
+- Show only system-profile by default?
+- How to handle machine-readability?
+- Provide stable API (e.g. `--porcelain`) for futureprooving output changes for
+scripts? (probably unnecessary, since usage in scripts probably limited)
+- Use `column` format for human readability? => Evaluation is eager
+- Use `pager` to avoid spamming terminal?
+- Is the order fine?
```
Comment (zimbatm) [inline rfcs/0086-nixos-generation-apis.md:187]: All of these are good questions. The best way to answer them is to play with an implementation to get a feel for it.

Targetting the interface for human-consumption first will allow more flexibility and to change the output as needed. Then when it's stable, add a --json output for machine consumption.

**Review by zimbatm — APPROVED** [review]
This doesn't look controversial to me and could be implemented directly.

File: rfcs/0086-nixos-generation-apis.md (hunk 2 above)
Comment (miallo) [reply rfcs/0086-nixos-generation-apis.md:141]: Yes, it sort of is independent, but the current implementation will just print the default value for any unrecognised flag and so in the implementation of list-generations I had to [implement a workaraound](https://github.com/NixOS/nixpkgs/blob/4ebca19b171dd5d0d0660ebfd9c413d24a70374e/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L448-L455) that ignores the output if it is the default. I agree -  I probably should have put this in another RFC and kept the default behaviour here...

File: rfcs/0086-nixos-generation-apis.md (hunk 1 above)
Comment (miallo) [reply rfcs/0086-nixos-generation-apis.md:17]: I thought about this and my thoughts were, to keep the interface as similar to the `--rollback` as possible. Also you could just as well `test`/ a generation. I thought about making `switch` the default (overridable) action but my main doubts came when @edolstra said
> There is an issue somewhere about replacing `nixos-rebuild` with a `nixos` command

and then I think `nixos --generation <n>` would be missing a verb ^^ (but maybe that is just personal preference...)

In general I got a bit unsatisfied with the name `switch` when I realised that the default behaviour is really doing a `build-and-switch` and here we really just want to switch to something already existing. This drove me a bit crazy when I learned about profiles while implementing this and I didn't know how to distinguish "switching to an existing profile" vs "creating a new generation of a profile". Actually I am thinking now, that the `--generation` flag is the only current possibility to just switch profiles - but I am probably overlooking something...

**Comment by miallo** [comment]
> This looks like a straight forward feature to me. Is there anything in particular that you would want to discuss/solve with having this as a RFC?

It is mostly about the [unresolved questions](https://github.com/miallo/rfcs/blob/86/rfcs/0086-nixos-generation-apis.md#list-generations-4) for which I wanted to get some comments :)


> ignore this message. I reread the RFC.

I 100% agree - the title I chose isn't the best...

File: rfcs/0086-nixos-generation-apis.md (hunk 1 above)
Comment (zimbatm) [reply rfcs/0086-nixos-generation-apis.md:17]: My bad. I forgot that the options are defined on the top-level.

In some sense, `--generation <n>` is a more general version of `--rollback` so I guess both flags should be mutually-exclusive.

Comment (miallo) [reply rfcs/0086-nixos-generation-apis.md:17]: I didn't enforce it with error messages, but the `--generation` is in an `else`-clause that doesn't get executed if both are given. Should I throw errors on all incompatible / useless combinations of commands? Right now they just don't do anything (i.e. `nixos-rebuild list-generations --rollback` will just list them and ignore the rest).

**Comment by miallo** [comment]
I was also a bit unsure if this should be an RFC. Since I am relatively new to the Nix-environment I wanted more input from more experienced people. An example: I didn't know about profiles and only considered the system one in my first implementation...

Did I understand the [rfc steering comittee meeting summary](https://github.com/NixOS/rfc-steering-committee/issues/56) right that I should close this issue?

**Pull request closed without merging**